test/hotspot/gtest/runtime/test_os.cpp

```diff
@@ -85,29 +85,29 @@ TEST_VM(os, page_size_for_region_alignment) {
 TEST_VM(os, page_size_for_region_unaligned) {
   if (UseLargePages) {
     // Given exact page size, should return that page size.
-    for (size_t i = 0; os::_page_sizes[i] != 0; i++) {
-      size_t expected = os::_page_sizes[i];
-      size_t actual = os::page_size_for_region_unaligned(expected, 1);
-      ASSERT_EQ(expected, actual);
+    for (size_t s = os::page_sizes().largest(); s != 0; s = os::page_sizes().next_smaller(s)) {
+      size_t actual = os::page_size_for_region_unaligned(s, 1);
+      ASSERT_EQ(s, actual);
     }
 
     // Given slightly larger size than a page size, return the page size.
-    for (size_t i = 0; os::_page_sizes[i] != 0; i++) {
-      size_t expected = os::_page_sizes[i];
-      size_t actual = os::page_size_for_region_unaligned(expected + 17, 1);
-      ASSERT_EQ(expected, actual);
+    for (size_t s = os::page_sizes().largest(); s != 0; s = os::page_sizes().next_smaller(s)) {
+      size_t actual = os::page_size_for_region_unaligned(s + 17, 1);
+      ASSERT_EQ(s, actual);
     }
 
     // Given a slightly smaller size than a page size,
     // return the next smaller page size.
-    if (os::_page_sizes[1] > os::_page_sizes[0]) {
-      size_t expected = os::_page_sizes[0];
-      size_t actual = os::page_size_for_region_unaligned(os::_page_sizes[1] - 17, 1);
-      ASSERT_EQ(actual, expected);
+    for (size_t s = os::page_sizes().largest(); s != 0; s = os::page_sizes().next_smaller(s)) {
+      const size_t expected = os::page_sizes().next_smaller(s);
+      if (expected != 0) {
+        size_t actual = os::page_size_for_region_unaligned(expected - 17, 1);
```

Comment:
@tstuefe should **expected** here be **s** ?

It seems like we are trying to compare the next smaller page size of S, with a slightly smaller size of the size S.

`os::page_size_for_region_unaligned(expected - 17, 1);`
vs.
`os::page_size_for_region_unaligned(s- 17, 1);`

Running these tests with 2 largepage sizes (3 total sizes) fails, however default of 2 page_sizes passes (4k and 2m or 1g):
(./hotspot/variant-server/libjvm/gtest/gtestLauncher -jdk:./images/jdk -Xms2G -Xmx2G -XX:+UseLargePages -XX:LargePageSizeInBytes=2M)
> [----------] 17 tests from os                                                                                                                                                                          [ RUN      ] os.page_size_for_region_vm                                                                                                                                                                [       OK ] os.page_size_for_region_vm (0 ms)                                                                                                                                                         [ RUN      ] os.page_size_for_region_aligned_vm                                                                                                                                                        [       OK ] os.page_size_for_region_aligned_vm (0 ms)                                                                                                                                                 [ RUN      ] os.page_size_for_region_alignment_vm                                                                                                                                                      [       OK ] os.page_size_for_region_alignment_vm (0 ms)                                                                                                                                               [ RUN      ] os.page_size_for_region_unaligned_vm                                                                                                                                                      test/hotspot/gtest/runtime/test_os.cpp:106: Failure                                                                                                                                                    Expected equality of these values:                                                                                                                                                                       actual                                                                                                                                                                                                   Which is: 4096                                                                                                                                                                                       expected                                                                                                                                                                                                 Which is: 2097152                                                                                                                                                                                  [  FAILED  ] os.page_size_for_region_unaligned_vm (0 ms)

Reply:
See https://github.com/openjdk/jdk/pull/1719 for fix.

Reply:
Yes, you are right, it should. I filed JDK-8257989. Feel free to fix it.

Thanks, Thomas

Reply:
Oh, I see you already have a PR open, I reply there.

Reply:
Fixed in #1719. Thanks Thomas.